rfcs/sdk-js/RFC.md

```diff
@@ -0,0 +1,544 @@
+
+```ts
+i18n("welcome", { name: "Inlang" })
+```
+
+### alternate links
+
+[Alternate links](https://developers.google.com/search/docs/specialty/international/localized-versions?hl=de) are an important way to tell search engines that this page exists in multiple languages.
+
+If we don't care about localized routes, we can auto-generate them. `inlang` knows about all languages a project can have. An url will just slightly change depending on the [`language detection`](#language-detection) strategy.
+
+If developers want to customize the behavior, they can manually call a `setAlternateLink` function.
+
+The adapter should auto-inject this metadata into the rendered HTML.
+
+### configuration
+
+Internationalization can be done in many different ways. Some things will work across all strategies e.g. how to display a translated `Message` on screen (see [lookup function](#lookup-function)). But other things need to be configured because each company/team may have it's own opinion how on things should work.
+
+That's why we need some kind of configuration options that an adapter can parse and follow to output different behavior.
+
+Having those options in the `inlang.config.js` means that the configuration can be moved to a different repository with any other framework and (as long as there exists an SDK for that) we will have the same i18n experience without setting up anything other than the adapter.
+
+Configuration options can be:
+
+#### language detection
+
+There are many ways how to detect the language that should be used:
+
+- `rootSlug` (default): e.g. `www.inlang.com/de/docs` with option to leave out the reference language
+- `TLD`: e.g. `www.inlang.de/docs`
+- `subdomain`: e.g. `de.inlang.com/docs`
+- `queryParameter`: e.g. `www.inlang.com/docs?lang=de`
+- auto-detection
+  - `acceptLanguageHeader`: e.g. `'accept-language: en;q=0.8, de;q=0.7, *;q=0.5'`
+  - `cookie`: reads a cookie value
+  - `header`: e.g. read the Cloudflare `CF-IPCountry` header
+  - `navigator`: for SPAs to get the language information client-side
+  - others not mentioned here, can be easily added
+- other functions that should be left to user land and therefore will not be part of the configuration
+  - get country information from the user's IP address
+  - get language information from the user object stored in DB
+  - custom; some things we didn't think of
+
+There is the need to support multiple strategies. If the first one does not match, the next strategies will be tried until a match is found. If nothing matches, the `referenceLanguage` will be used.
+
+Setting up auto detection is optional. Developers can also just call the `changeLanguage` function directly.
+
+Each strategy will need a special framework agnostic implementation, to pass the necessary information to those detection functions.
+
+For the first version we probably won't need all of the strategies.
+
+> What detection strategies do we want to support?
```

Comment:
I would vote for `rootSlug`, `acceptLanguageHeader` and `navigator` since it is possible to cover SPAs, MPAs and content websites.